bpf/bpf_lxc.c

```diff
@@ -1036,6 +1077,21 @@ int tail_ipv6_policy(struct __ctx_buff *ctx)
 
 	/* Store meta: essential for proxy ingress, see bpf_host.c */
 	ctx_store_meta(ctx, CB_PROXY_MAGIC, ctx->mark);
+
+#ifdef ENABLE_CUSTOM_CALLS
+	/* Make sure we skip the tail call when the packet is being redirected
+	 * to a L7 proxy, to avoid running the custom program twice on the
+	 * incoming packet (before redirecting, and on the way back from the
+	 * proxy).
+	 */
+	if (ret != CTX_ACT_TX &&
```

Comment:
I think that the `ret != CTX_ACT_TX` is making an assumption about the way we forward to the proxy here when really we should just be checking the condition that occurs on line 1072. At that earlier point in the function, `ret == POLICY_ACT_PROXY_REDIRECT` means that we are definitely sending the traffic to the proxy.

The assumption is that to forward to the proxy, a device redirect is required. That is not the case if `from_host` is true in this context. The `CTX_ACT_TX` here means we will redirect to another device to continue processing, but if for instance we are performing a policy lookup at source device for host->pod on same node, then I don't think that we will set the `ret` to `CTX_ACT_TX` in that case when forwarding traffic to the proxy. So we'll double-trace that particular case.

(Presumably the same for IPv4).